junit-common/src/main/java/org/jboss/weld/junit/common/MockBean.java

```diff
@@ -0,0 +1,387 @@
+/*
+ * JBoss, Home of Professional Open Source
+ * Copyright 2017, Red Hat, Inc., and individual contributors
+ * by the @authors tag. See the copyright.txt in the distribution for a
+ * full listing of individual contributors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jboss.weld.junit.common;
+
+import java.lang.annotation.Annotation;
+import java.lang.reflect.Type;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import javax.enterprise.context.Dependent;
+import javax.enterprise.context.spi.CreationalContext;
+import javax.enterprise.inject.Any;
+import javax.enterprise.inject.Default;
+import javax.enterprise.inject.spi.Bean;
+import javax.enterprise.inject.spi.InjectionPoint;
+import javax.enterprise.inject.spi.PassivationCapable;
+import javax.enterprise.util.AnnotationLiteral;
+
+/**
+ * This custom {@link Bean} implementation is useful for mocking.
+ * <p>
+ * A new instance is usually created through a {@link Builder} (see also {@link #builder()}) and then passed to the
+ * {@link WeldInitiator.Builder#addBeans(Bean...)} method.
+ * </p>
+ *
+ * @author Martin Kouba
+ *
+ * @param <T>
+ * @see WeldInitiator.Builder#addBean(Bean)
+ * @since 1.1
+ */
+public class MockBean<T> implements Bean<T>, PassivationCapable {
```

Comment:
This is just copied over, I am not sure we want an `abstract` version of this bean?
One of the arguments against is that junit 5 == java 8 which means we can use lambdas instead of the interfaces added here. Besides it gets clumsy with the static builder methods.

Reply:
Abstract version is not needed. 

> One of the arguments against is that junit 5 == java 8 which means we can use lambdas instead of the interfaces added here.

`CreateFunction` could be used as lambda too, `@FunctionalInterface` is just a marker annotation.

> Besides it gets clumsy with the static builder methods.

What do you mean?

Reply:
I mean that those static methods would still have to be on concrete impl, but nevermind.
So, if we don't go for abstract version, that means we will have to have this bean in both modules so that users don't need to depend on the "common" module, or import classes from there.

Reply:
Why? I suppose `common` module is a dependency of `junit4` and `junit5` modules.

Reply:
Yes but then in your test class you are going to have (amongst other things):
`import org.jboss.weld.junit.common.MockBean;`
`import org.jboss.weld.junit5.WeldInitiator;`
`import org.jboss.weld.junit5.WeldJunit5Extension;`

We need to have the packages different (J9) but this looks inconvenient, you would expect to have all the needed classes in the same package, wouldn't you?

Reply:
No, I wouldn't. I don't think it's a problem. We can also rename `org.jboss.weld.junit.common` to `org.jboss.weld.junit`...